src/mod_antispam_rtbl.erl

```diff
@@ -22,80 +22,94 @@
 %%% 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 %%%
 %%%----------------------------------------------------------------------
+
+%%| definitions
+%% @format-begin
+
 -module(mod_antispam_rtbl).
+
 -author('stefan@strigler.de').
 
 -include_lib("xmpp/include/xmpp.hrl").
+
 -include("logger.hrl").
 -include("mod_antispam.hrl").
 
 -define(SERVICE_MODULE, mod_antispam).
 -define(SERVICE_JID_PREFIX, "rtbl-").
 
--export([parse_blocked_domains/1,
-	 parse_pubsub_event/1,
-	 pubsub_event_handler/1,
-	 request_blocked_domains/3,
-	 subscribe/3,
-	 unsubscribe/3]).
+-export([request_blocked_domains/2, cancel_timers/1, handle_iq_reply/3, unsubscribe/2,
+         get_service/3]).
+%% Hooks
+-export([add_hook/1, delete_hook/1, pubsub_event_handler/1]).
 
-%% @format-begin
+%%--------------------------------------------------------------------
+%%| Route IQs
 
-subscribe(RTBLHost, RTBLDomainsNode, From) ->
+subscribe(#rtbl_service{host = Host, node = Node}, From) ->
     FromJID = service_jid(From),
     SubIQ =
         #iq{type = set,
-            to = jid:make(RTBLHost),
+            to = jid:make(Host),
             from = FromJID,
-            sub_els = [#pubsub{subscribe = #ps_subscribe{jid = FromJID, node = RTBLDomainsNode}}]},
+            sub_els = [#pubsub{subscribe = #ps_subscribe{jid = FromJID, node = Node}}]},
     ?DEBUG("Sending subscription request:~n~p", [xmpp:encode(SubIQ)]),
     ejabberd_router:route_iq(SubIQ, subscribe_result, self()).
 
--spec unsubscribe(binary() | none, binary(), binary()) -> ok.
-unsubscribe(none, _PSNode, _From) ->
-    ok;
-unsubscribe(RTBLHost, RTBLDomainsNode, From) ->
+-spec service_jid(binary()) -> jid().
+service_jid(Host) ->
+    jid:make(<<>>, Host, <<?SERVICE_JID_PREFIX, (ejabberd_cluster:node_id())/binary>>).
+
+-spec unsubscribe([rtbl_service()] | rtbl_service(), binary()) -> ok.
+unsubscribe(Services, From) when is_list(Services) ->
+    [unsubscribe(Service, From) || Service <- Services];
+unsubscribe(#rtbl_service{host = Host, node = Node}, From) ->
     FromJID = jid:make(From),
     SubIQ =
         #iq{type = set,
-            to = jid:make(RTBLHost),
+            to = jid:make(Host),
             from = FromJID,
-            sub_els =
-                [#pubsub{unsubscribe = #ps_unsubscribe{jid = FromJID, node = RTBLDomainsNode}}]},
+            sub_els = [#pubsub{unsubscribe = #ps_unsubscribe{jid = FromJID, node = Node}}]},
     ejabberd_router:route_iq(SubIQ, unsubscribe_result, self()).
 
--spec request_blocked_domains(binary() | none, binary(), binary()) -> ok.
-request_blocked_domains(none, _PSNode, _From) ->
-    ok;
-request_blocked_domains(RTBLHost, RTBLDomainsNode, From) ->
+-spec request_blocked_domains([rtbl_service()] | rtbl_service(), binary()) -> any().
+request_blocked_domains(Services, From) when is_list(Services) ->
+    [request_blocked_domains(Service, From) || Service <- Services];
+request_blocked_domains(#rtbl_service{host = Host, node = Node}, From) ->
```

Comment:
Since you‘re not making a differentiation here, you‘d also request „blocked domains“ from muc_bans_sha256 nodes. This might not be problem here except for the naming of things, but I haven‘t found in your code where you‘d handle the response differently either.

Since there are no tests for `mod_muc_rtbl`, have you tested this manually, that this is still working? Could you add automated tests for `muc_bans_sha256` please?

Reply:
This is the data flow:
- `request_blocked_domains` sends queries to all the configured RTBL services.
- `parse_blocked_domains` manages all responses equally, regardless if they are JID or SHA. This means blocked_domains contains a mix of JIDs and SHAs.
- Finally, in `handle_call({is_blocked_domain, Domain},`, if the inspected JID is not found in blocked_domains, it searches for its SHA too.

An optimization would be to store always SHA, either because it was provided or because we calculate at storage time. Would be nice to store also JID in another field when it's known. Then search directly by SHA.